contracts/implementations/nodes/AftermarketDevice.sol

```diff
@@ -246,6 +246,86 @@ contract AftermarketDevice is
         emit AftermarketDeviceClaimed(aftermarketDeviceNode, owner);
     }
 
+    /// @notice Pairs an aftermarket device with a vehicle through a metatransaction.
+    /// Both vehicle and AD owners can pair.
+    /// The vehicle/AD owner signs a typed structured (EIP-712) message in advance and submits to be verified
+    /// @dev Caller must have the admin role
+    /// @param aftermarketDeviceNode Aftermarket device node id
+    /// @param vehicleNode Vehicle node id
+    /// @param vehicleOwnerSig Vehicle owner signature hash
+    /// @param aftermarketDeviceSig Aftermarket Device's signature hash
+    function pairAftermarketDeviceSign(
+        uint256 aftermarketDeviceNode,
+        uint256 vehicleNode,
+        bytes calldata aftermarketDeviceSig,
+        bytes calldata vehicleOwnerSig
+    ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        MapperStorage.Storage storage ms = MapperStorage.getStorage();
+        bytes32 message = keccak256(
+            abi.encode(PAIR_TYPEHASH, aftermarketDeviceNode, vehicleNode)
+        );
+        address vehicleIdProxyAddress = VehicleStorage
+            .getStorage()
+            .idProxyAddress;
+        address adIdProxyAddress = AftermarketDeviceStorage
+            .getStorage()
+            .idProxyAddress;
+
+        require(
+            INFT(vehicleIdProxyAddress).exists(vehicleNode),
+            "Invalid vehicle node"
+        );
+
+        require(
+            INFT(adIdProxyAddress).exists(aftermarketDeviceNode),
+            "Invalid AD node"
+        );
+        require(
+            AftermarketDeviceStorage.getStorage().deviceClaimed[
+                aftermarketDeviceNode
+            ],
+            "AD must be claimed"
+        );
+        require(
+            ms.links[vehicleIdProxyAddress][vehicleNode] == 0,
+            "Vehicle already paired"
+        );
+        require(
+            ms.links[adIdProxyAddress][aftermarketDeviceNode] == 0,
+            "AD already paired"
+        );
+        require(
+            Eip712CheckerInternal._verifySignature(
+                AftermarketDeviceStorage.getStorage().nodeIdToDeviceAddress[
+                    aftermarketDeviceNode
+                ],
+                message,
+                aftermarketDeviceSig
+            ),
+            "Invalid signature"
+        );
+
+        address adOwner = INFT(vehicleIdProxyAddress).ownerOf(vehicleNode);
+
+        require(
+            Eip712CheckerInternal._verifySignature(
+                adOwner,
+                message,
+                vehicleOwnerSig
+            ),
+            "Invalid signature"
+        );
+
+        ms.links[vehicleIdProxyAddress][vehicleNode] = aftermarketDeviceNode;
+        ms.links[adIdProxyAddress][aftermarketDeviceNode] = vehicleNode;
+
+        emit AftermarketDevicePaired(
+            aftermarketDeviceNode,
+            vehicleNode,
+            adOwner
```

Comment:
But maybe this should still be the aftermarket owner, for consistency? Or does it always hold the person in the loop? I don't know.

Reply:
Yep. I think makes more sense keeping the AD owner. Don't have a strong opinion why, just a feeling